core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
 
-  def connect(mesosHostName: String): Future[(MesosConnection, Flow[SpecEvent, StateEvent, NotUsed])] = {
-    val flow = Flow.fromGraph {
-      GraphDSL.create(Scheduler.unconnectedGraph, FakeMesos.flow)((_, _) => NotUsed) { implicit builder =>
-        { (graph, mockMesos) =>
+  type StateOutput = (StateSnapshot, Source[StateEvent, Any])
+
+  case class MesosConnection(mesosHostName: String, mesosCallFactory: MesosCalls, frameworkId: FrameworkID, schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])
+
+  // TODO - provide abstraction for FrameworkID persistence, recovery, etc.; we'll want to exclude the frameworkId
+  // TODO - This layer should provide a cancellation mechanism and an auto-reconnect mechanism
+  def connect(settings: MesosClientSettings, frameworkInfo: FrameworkInfo)(
+      implicit actorSystem: ActorSystem,
+      materializer: ActorMaterializer): Future[MesosConnection] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    MesosClient(settings, frameworkInfo).runWith(Sink.head).map { client =>
+      val flow = mesosConnectedGraph(client.calls, Flow.fromSinkAndSource(client.mesosSink, client.mesosSource))
+      MesosConnection(settings.master, client.calls, client.frameworkId, flow)
+    }
+  }
+
+  def mesosConnectedGraph(mesosCallFactory: MesosCalls, mesosFlow: Flow[MesosCall, MesosEvent, Any]): Flow[SpecInput, StateOutput, NotUsed] = {
+    Flow.fromGraph {
+      GraphDSL.create(unconnectedGraph(mesosCallFactory), mesosFlow)((_, _) => NotUsed) { implicit builder =>
+        { (graph, mesos) =>
           import GraphDSL.Implicits._
 
-          mockMesos ~> graph.in2
-          graph.out2 ~> mockMesos
+          mesos ~> graph.in2
+          graph.out2 ~> mesos
 
           FlowShape(graph.in1, graph.out1)
         }
       }
     }
-    Future.successful(MesosConnection(mesosHostName) -> flow)
   }
 
-  /*
-   */
-  def unconnectedGraph: BidiFlow[SpecEvent, StateEvent, Mesos.Event, Mesos.Call, NotUsed] = {
+  /**
+    * We expess an interface of receiving a snapshot, followed by a series of events, to guide consumers of USI down a
+    * proper implementation path. A SpecSnapshot should be the very first thing that the USI Scheduler receives from the
+    * Framework implementation.
+    *
+    * However, for convenience in working with streams, internally we deal with a single stream of SpecEvents.
+    */
+  private val specInputFlatteningFlow: Flow[SpecInput, SpecEvent, NotUsed] = Flow[SpecInput].flatMapConcat { case (snapshot, rest) =>
+    rest.prepend(Source.single(snapshot))
+  }
+
+  private val stateOutputBreakoutFlow: Flow[StateEvent, StateOutput, NotUsed] = Flow[StateEvent].prefixAndTail(0).map { case (_, stateEvents) =>
+    // TODO (DCOS-47476) use actual prefixAndTail and expect first event to be a Snapshot
+    val stateUpdates = stateEvents.map {
+      case c: StateUpdated => c
+      case _ => ???
+    }
+    (StateSnapshot.empty, stateUpdates)
```

Comment:
The other way around (see my comment above), I'm not sure what difference does it make for the orchestration layer to receive `StateSnapshot` as opposed to a stream of `StateUpdated`.

Reply:
I believe it's just enforcing the order of Snapshot always being first. You need the snapshot to finish initialization of your orchestration layer

Reply:
Discussed